langgenius/dify
# feat: app rate limit

# Description

**Feature**
The "apps" table now includes a new field, "max_active_requests," used to control the maximum concurrency of applications. Additionally, the configuration now features "APP_MAX_ACTIVE_REQUESTS" to set the default maximum concurrency value for applications, with 0 indicating no limit.

**Fixes**
Applications of the agent type couldn't stream return results when xinference was used as the model provider.


## Type of Change

Please delete options that are not relevant.


# How Has This Been Tested?

**Here are the testing steps for the feature:**
1. In the "apps" table, find an application and set "max_active_requests" to 1.
2. Add a Redis record of type hash with the key dify:rate_limit:<app_id>:active_requests and a data entry where the key is arbitrary and the value is the current timestamp (time.time())
![image](https://github.com/langgenius/dify/assets/11787325/e52a34c2-e078-4918-b555-366f2d3abd9a)
4. Start the dify api module
5. Use the application to engage in a conversation.
These steps will help verify the functionality of limiting maximum concurrent requests for the application.
![image](https://github.com/langgenius/dify/assets/11787325/d5c58357-eda7-44eb-b9ef-62884423d8fb)


**Here are the testing steps for the bug:**
1. Create an application with the type "agent."
2. Set the model to be provided by xinference.
3. Engage in a conversation using the application.

**Expected outcome:**
The application should stream the results back as expected. This test will help verify that the bug related to streaming return results has been resolved.


# Suggested Checklist:
